MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1774 @@
+
+      track_set tracks[2];
+      tracks[0] = vertex_track_set(*v[0]);
+
+      bool merge = false;
+      for (v[1] = v[0] + 1; v[1] != vertices->end(); ++v[1]) {
+        if (vertices->size() >= 2 && v[0]->nTracks() >= 2 && v[1]->nTracks() >= 2) {
+
+          tracks[1] = vertex_track_set(*v[1]);
+
+          Measurement1D v_dist = vertex_dist_2d.distance(*v[0], *v[1]);
+
+          Measurement1D dBV0_Meas1D = vertex_dist_2d.distance(*v[0], fake_bs_vtx);
+          double dBV0 = dBV0_Meas1D.value();
+
+          Measurement1D dBV1_Meas1D = vertex_dist_2d.distance(*v[1], fake_bs_vtx);
+          double dBV1 = dBV1_Meas1D.value();
+
+          double v0x = v[0]->x() - bsx;
+          double v0y = v[0]->y() - bsy;
+
+          double phi0 = atan2(v0y, v0x);
+
+          double v1x = v[1]->x() - bsx;
+          double v1y = v[1]->y() - bsy;
+
+          double phi1 = atan2(v1y, v1x);
+
+          if (fabs(reco::deltaPhi(phi0, phi1)) < 0.5 && v_dist.value() < 0.0300 && dBV0 > 0.0100 && dBV1 > 0.0100) {
+            track_set tracks_to_fit;
+            for (int i = 0; i < 2; ++i)
+              for (auto tk : tracks[i])
+                tracks_to_fit.insert(tk);
+            std::vector<reco::TransientTrack> ttks;
+            for (auto tk : tracks_to_fit)
+              ttks.push_back(seed_tracks[seed_track_ref_map[tk]]);
+
+            if (investigate_merged_vertices) {
+              std::vector<TransientVertex> tv(1, kv_reco->vertex(ttks));
+              potential_merged_vertices.push_back(reco::Vertex(tv[0]));
+              //std::cout << "ntrack in potental merged: " << potential_merged_vertices.back().nTracks() << std::endl;
+            }
+            reco::VertexCollection merged_vertices;
+            for (const TransientVertex& tv : kv_reco_dropin(ttks))
+              merged_vertices.push_back(reco::Vertex(tv));
+
+            if (merged_vertices.size() == 1 && vertex_track_set(merged_vertices[0], 0) == tracks_to_fit) {
+
+              merge_pair_count += 1;
+              merge = true;
+
+              if (verbose) {
+                printf(" sv2ddist between a merging pair is %7.3f \n", v_dist.value());
+                printf(" |dPhi(vtx0,vtx1) between a merging pair is %4.3f \n", fabs(reco::deltaPhi(phi0, phi1)));
+                printf(" # of tracks per vtx0 is %u \n", v[0]->nTracks());
+                printf(" # of tracks per vtx1 is %u \n", v[1]->nTracks());
+                printf(" ---------------- merge the two vertices if chi2/dof < 8 ----------------- \n");
+                printf(" # of tracks per a new merged vertex is %u \n", merged_vertices[0].nTracks());
+              }
+
+              v[1] = vertices->erase(v[1]) - 1;
+              *v[0] = reco::Vertex(merged_vertices[0]); // ok to use v[0] after the erase(v[1]) because v[0] is by construction before v[1]
+
+            }
+          }
+        }
+      }
+
+      if (merge)
+        v[0] = vertices->begin() - 1;
+    }

```

Comment:
Let's add a comment here about what is happening. I think it is that we go through all possible pairings w/ v[0] and perform our merges. And then at the end of the v[0] loop, this points the iterator right before begin(), so that we go through all possible (new) combinations again, in case there is now a valid merge.

Does that summarize it correctly? We might be able to come up with a super short description to add as a comment here, if this is accurate.

Reply:
Yes, that is correct. I can try to add a short description here.

Reply:
Done.